### PR TITLE
Eco-Credit Module Proof of Concept

### PR DESCRIPTION
**Pull request by aaronc**
ref: #78, #117 

Implements the `x/ecocredit` module as defined in #125. Read that PR description first for context. Module format and tests are experimental and follow what was done in #126. 

Proto breaking changes from #125:
* `issuer_address` reverted to `issuer` for consistency (feel free to improve docs where account address isn't clear)
* `EventReceive` updated based on https://github.com/regen-network/regen-ledger/pull/125/files#r517647028

---

Feel free to submit a Draft Pull Request as soon as you start working on something. Before you mark your PR as Ready For Review, make sure you've checked off the following boxes or indicated N/A (not applicable):
- [ ] linked to a relevant issue or have written a clear description of the issue in the PR
- [ ] wrote or updated tests
- [ ] wrote or updated documentation
- [ ] added an appropriate entry in `CHANGELOG.md` following https://keepachangelog.com

Thanks!

### CODE REVIEW

**Comment by robert-zaremba** [comment]
Is this ready for review? This PR includes the Regen Ledger commits, which makes it hard to focus on the problem (Credit Module)

**Comment by aaronc** [comment]
> Is this ready for review? This PR includes the Regen Ledger commits, which makes it hard to focus on the problem (Credit Module)

yeah it was messy, all cleaned up now

**Comment by aaronc** [comment]
If someone can pick this up in the AM that would be great. @clevinson didn't have time to review tonight. I consider it basically done, tests are passing. So review, make and necessary changes, merge when you feel it's ready and keep going.

File: x/ecocredit/server/helpers.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	err := binary.Write(buf, binary.LittleEndian, value)
+	if err != nil {
+		return err
+	}
+
+	store.Set(key, buf.Bytes())
+	return nil
+}
+
+func getUint32(store sdk.KVStore, key []byte) (uint32, error) {
```
Comment (robert-zaremba) [inline x/ecocredit/server/helpers.go:80]: Shouldn't we use `uint64`?

**Review by amaurymartiny — APPROVED** [review]
utACK. I didn't do a thorough review of math.go, but msgs, queries and tests lgtm.

**Review by blushi — APPROVED** [review]
LGTM

**Comment by robert-zaremba** [comment]
Sorry for not looking at it earlier today. When I looked at it [previously](https://github.com/regen-network/regen-ledger/pull/119#issuecomment-721984897) it was hard to digest what it does as it included all new regen codebase commits.

**Review by robert-zaremba — APPROVED** [review]
In the future we should move the `math` library out and rename it to `decimals` or something similar (dec, decim...)

File: x/ecocredit/server/helpers.go (hunk 1 above)
Comment (aaronc) [reply x/ecocredit/server/helpers.go:80]: Well, decimal places is just uint32 which seems like plenty

**Comment by aaronc** [comment]
All good @robert-zaremba. Have you seen the various discussions about decimals in the SDK?

**Comment by robert-zaremba** [comment]
> All good @robert-zaremba. Have you seen the various discussions about decimals in the SDK?

I've seen the one you pointed that `sdk.Dec` doesn't work. I've created an issue (https://github.com/regen-network/regen-ledger/issues/149) to follow up on that and linked it with https://github.com/cosmos/cosmos-sdk/issues/7113

**Pull request merged**